### PR TITLE
Add support for HuggingFace Datasets

### PR DESCRIPTION
**Pull request by ernestum**
## Description

As per #651 we want to start storing trajectories in HuggingFace Datasets

This PR adds support for HuggingFace Datasets and moves `imitation.data.types.load/save` to `imitation.data.seralize.load/save`. Datasets are stored in the HuggingFace format by default.

The conversion script is adapted accordingly and I re-structured the tests.

I also moved the loading/saving functionality to `imitation.data.serialize` instead of `imitation.data.types` to mimic the layout of `imitation.policies`.

## Design Choices
I introduced a new class ([`imitation.data.huggingface_datasets_conversion.TrajectoryDatasetSequence`](https://github.com/HumanCompatibleAI/imitation/blob/6e82c9a51f0245021c65f3d098e77db4e99654ba/src/imitation/data/huggingface_datasets_conversion.py#L12)) that presents a HuggingFace dataset as a sequence of `imitation.data.types.Trajectory`. In its [`__getitem__` ](https://github.com/HumanCompatibleAI/imitation/blob/6e82c9a51f0245021c65f3d098e77db4e99654ba/src/imitation/data/huggingface_datasets_conversion.py#L34) method the dataset is queried and the result is converted to a trajectory (or slice of trajectories) ad-hoc.

I found that the "info" dicts were very heterogeneous in nature and could therefore not be mapped to some dataset layout. I decided to serialize the info dicts using `jsonpickle` and then just store a list of strings. Since decoding this is far more expensive than reading the other (potentially memory-mapped) features (observations, actions, terminals) and often we don't care about the "infos", I decided for a [lazy decoding wrapper](https://github.com/HumanCompatibleAI/imitation/blob/6e82c9a51f0245021c65f3d098e77db4e99654ba/src/imitation/data/huggingface_datasets_conversion.py#L60). It will only decode the info dicts when accessed and keep decoded dicts in an LRU cache to avoid decoding them multiple times.

## Testing

Old tests have been updated and split up into more atomic unit tests.

### CODE REVIEW

**Comment by ernestum** [comment]
@AdamGleave Right now the tests fail due to lack of git lfs support in our Docker image. Is there any reason we don't use git lfs? If not, I would add it according to this guide: https://naiyer.dev/post/2020/09/05/using-git-lfs-in-ci/

**Comment by ernestum** [comment]
Not sure what is going on with the coverage reports again :shrug:

**Comment by AdamGleave** [comment]
@Rocamonde can you review this please?

File: src/imitation/algorithms/dagger.py
Comment (Rocamonde) [inline src/imitation/algorithms/dagger.py:24]: Why `import imitation.data.serialize` and also `from imitation.data import serialize`? Better to keep things consistent.

File: tests/scripts/test_scripts.py
Comment (Rocamonde) [inline tests/scripts/test_scripts.py:1018]: Are we doing GWT for all tests now?

File: src/imitation/data/huggingface_datasets_conversion.py
Comment (Rocamonde) [inline src/imitation/data/huggingface_datasets_conversion.py:83]: I guess this is beyond the scope of reviewing the PR, but why is this better than pickling a dataclass? A compelling case for this would be very relevant for other current projects.

File: src/imitation/data/serialize.py
Comment (Rocamonde) [inline src/imitation/data/serialize.py:11]: Hmmm... Maybe have a shorter module name in the first place?

File: src/imitation/algorithms/dagger.py
Comment (ernestum) [reply src/imitation/algorithms/dagger.py:24]: Good point. Fixed it.

File: tests/scripts/test_scripts.py
Comment (ernestum) [reply tests/scripts/test_scripts.py:1018]: Yes.

I think it makes tests a lot easier to grasp when reading them again.
It also pushes us to only test one concept at a time in each test. This makes it a lot easier to find the root cause when a test fails.

File: src/imitation/data/huggingface_datasets_conversion.py
Comment (ernestum) [reply src/imitation/data/huggingface_datasets_conversion.py:83]: Valid question I think! For me there are the following points:

- It is generally not a good idea to serialize classes for persistence. Whenever you change the class implementation, you might (silently) loose the ability to load older files. I had this happen when loading SB3 models trained with old gym versions (with gym changing implementations with no regard to backwards compatibility)
- The pickle format changes from time to time and sometimes I have to mess with pickle versions.
- The HuggingFace Datasets are backed by Apache Arrow. This is mostly relevant in case our datasets grow. It allows to stream datasets (from disk or from the internet) and does memory mapping for us. With the memory mapping we can load huge files that would never fit in RAM and let the OS do transparent caching for us.
- The HuggingFace datasets do not support pickled objects by default (probably due to the above reasons and to make users adhere to best practices). We could make it happen with a additional loading script next to the pickled classes but this would increase complexity and I really don't like storing code together with our datasets.

You can read more about the considerations in #651

File: src/imitation/data/serialize.py
Comment (ernestum) [reply src/imitation/data/serialize.py:11]: We could go with `hf_datasets_conversion`? I am happy about other suggestions but I would like the name to stay descriptive. I rather type some more letters now than wonder what a module was supposed to do later.

Comment (Rocamonde) [reply src/imitation/data/serialize.py:11]: I might call it huggingface_data_converter or something like that? I think having a very descriptive name but then renaming the import to a non-obvious abbreviation is probably just as bad for redability

Comment (ernestum) [reply src/imitation/data/serialize.py:11]: Hmm. I am not really happy with that name either. I would expect a class named `HuggingFaceDataConverter` in such a module.

When it is so hard to come up with a good name for a module that is often an indicator for bad architecture of the module. So I looked at its content again and the following came up:

When saving, we create a HF dataset just to write it to disk immediately. This involves a (shallow) copy of the data.

When loading, we have a wrapper that makes a HF dataset visible as a sequence of trajectories. Trajectory objects generated on-the-fly.

To make this symmetrical, I should have used the `Dataset.from_generator` method instead of the `Dataset.from_dict` method with a generator that constructs dicts from trajectories on-the-fly ([documentation here](https://huggingface.co/docs/datasets/v2.10.0/en/create_dataset#from-local-files)).
This is also the most memory-efficient way to create datasets and it would make it possible to stream sampled trajectories straight to disk.
I will push a version of this soon.

Concerning the naming of the module: I would propose to just call it `huggingface_utils` or even `hf_utils`. In symmetry to that we could refactor `imitation.policies.serialize` and pull out most of the HF-specific code in a `imitation.policies.huggingface_utils` module.

Comment (Rocamonde) [reply src/imitation/data/serialize.py:11]: That sounds like a good decision, thanks for taking a look at this! Huggingface_utils sounds good.

Comment (ernestum) [reply src/imitation/data/serialize.py]: Added it just now. Unfortunately this change made the save a lot slower (`test_types.test_save_trajectories` executes orders of magnitude slower). Maybe here @simoninithomas can give us some insight?

Comment (simoninithomas) [reply src/imitation/data/serialize.py]: So I asked the dataset team, that can explain the problem:
- `from_dict` writes in RAM
- `from_generator` writes on disk

Comment (ernestum) [reply src/imitation/data/serialize.py]: So should I better switch between `from_dict` and `from_generator` based on the size of the dataset (assuming that I know the number of elements coming out of the generator) or is that something, that you maybe implement on the datasets library end?

Comment (simoninithomas) [reply src/imitation/data/serialize.py]: If your dataset is a python dict, it takes up RAM and is therefore relatively small so you can load it with from_dict. But if you both load files one by one, then from_generator is better

**Comment by ernestum** [comment]
I finally decided to go without the `from_generator` function and only use it when we actually need it. Right now our datasets are small enough to fit in memory.

@Rocamonde could you give this another pass and then we can merge it?

**Comment by Rocamonde** [comment]
Sure! Will take a look tomorrow.

File: src/imitation/scripts/convert_trajs.py
Comment (Rocamonde) [inline src/imitation/scripts/convert_trajs.py:14]: In other places we're importing `from imitation.data import serialize`

File: src/imitation/data/serialize.py
Comment (Rocamonde) [inline src/imitation/data/serialize.py:1]: I understand you just moved the code into a separate file? Or are there any other additional changes? Hard to see based on the diff github is showing

Comment (Rocamonde) [reply src/imitation/data/serialize.py:1]: Otherwise all seems OK

File: tests/data/test_types.py
Comment (ernestum) [reply tests/data/test_types.py:46]: This is a test fixture for trajectories.
Before the test trajectories included info dicts whose keys were `int`s . This change ensures we only have dicts with `str` typed keys since
1. I don't think the infos dicts ever have any other key type except strings.
2. The current implementation relies on jsonpickle to serialize the info dicts. Jsonpickle would convert those int keys to strings anyways:
```python
>> jsonpickle.loads(jsonpickle.dumps({1: 1}))
{'1': 1}
```
This would make the tests fail that store a trajectory, load it again and compare the two. The original would have `int` keys while the loaded one would have `str` keys.

File: src/imitation/scripts/convert_trajs.py
Comment (ernestum) [reply src/imitation/scripts/convert_trajs.py:14]: Good point. Even found some more places where this is the case. I made sure it is always imported as `from imitation.data import serialize` or `from imitation import data` whenever we also need the `policies.serialize` module in the same file.

File: src/imitation/data/serialize.py
Comment (ernestum) [reply src/imitation/data/serialize.py:1]: I moved the serialization related functions to a separate file but also changed the save/load functions as they now use the huggingface backend. Sorry I also cant find a way to generate a decent diff for this.

**Review by Rocamonde — APPROVED** [review]
LGTM

File: src/imitation/scripts/convert_trajs.py
Comment (ernestum) [reply src/imitation/scripts/convert_trajs.py:53]: The main is not included, `update_traj_file_in_place` however is.

**Comment by AdamGleave** [comment]
Anything blocking merging now we have an LGTM? Is it just the code coverage?

**Comment by ernestum** [comment]
What is missing in coverage right now (as I interpret the codecov output)

- loading old versions of the `Trajectory` dataclass which did not have the `terminal` flag yet.
- specifying the log-level with something that can not be cast to an `int`

If you think we can live without that coverage then you can merge this please.

**Comment by AdamGleave** [comment]
Sorry for the slow turnaround -- yes, I think that coverage isn't essential, I have merged now.

**Pull request merged**
